### PR TITLE
Cache origin and domain normalization

### LINKED ISSUE #102: Stencil launch overhead
The launch overhead of gt4py stencils seems to be on the order of 10ms, which is a _lot_. The whole timestep of an atmospheric model should typically take O(100ms) and is launching O(100) stencils, if we are aiming for a throughput of 1 SYPD at 1 km resolution. The plot shows launch overheads of the same stencil running in a) plain python b) gt4py with numpy backend c) gt4py with gtx86 bakckend and d) Fortran.

image

### PR DESCRIPTION
**Pull request by jdahm**
## Description

This caches the stencil pre-computation normalization. The unit test stencil overhead is sped up by 25x on subsequent calls with the same origin and domain.

Builds on #537.

Resolves #102.

### CODE REVIEW

**Comment by jdahm** [comment]
@egparedes perhaps it should be a wrapper over StencilObject that is set with `cached=True` on the `gtscript.stencil` call?

File: src/gt4py/stencil_object.py
Comment (egparedes) [inline src/gt4py/stencil_object.py:44]: `hash()` should only be called **_once_** with an argument tuple containing all the information needed to create the key.

File: src/gt4py/stencil_object.py
Comment (egparedes) [inline src/gt4py/stencil_object.py:428]: I understand that checking if the current call matches the last one is the most common case, but I would add a dict here  `{cache_key: (origin, domain)}` (maybe limited in size?) to make sure that information from any previous call is reused, since the cost of computing the cache key is already much larger than a (small) dict lookup.

**Review by egparedes — CHANGES_REQUESTED** [review]
There is a lot of overlap with #537. I would wait until that one is merged before keep working on this. Besides that, I have a couple of comments regarding the implementation.

File: src/gt4py/stencil_object.py
Comment (jdahm) [inline src/gt4py/stencil_object.py:436]: Combine.

**Comment by egparedes** [comment]
#537 has been already merged. I can review this PR again once you have merged master into it and solved the conflicts.

File: src/gt4py/stencil_object.py
Comment (jdahm) [inline src/gt4py/stencil_object.py:130]: This is implemented at the expense of freezing the StencilObject class. Does anyone see this as a problem? We could potentially do this in a wrapper class a la FrozenStencil.

File: src/gt4py/stencil_object.py
Comment (egparedes) [inline src/gt4py/stencil_object.py:489]:
```suggestion
            domain, origin = type(self)._domain_origin_cache[cache_key]
```

**Review by egparedes — CHANGES_REQUESTED** [review]
I think there are some conceptual and implementation errors in the PR and thus I suggest changes to overcome them.
Most of the errors should have been caught by proper tests, so I suggest to also add new tests for the new functionality (showing that it works with and without cache hit, and that there is performance benefit for the cache hits).

**Comment by egparedes** [comment]
> @egparedes perhaps it should be a wrapper over StencilObject that is set with `cached=True` on the `gtscript.stencil` call?

Ooops, I had missed this comment, @jdahm. I like the idea and it shouldn't be hard to implement. Maybe we can fix first the current issues and then refactor the code following your idea.

File: src/gt4py/stencil_object.py
Comment (jdahm) [reply src/gt4py/stencil_object.py:40]: Yes, it seems like the storage implementation forces it to be a `Tuple[int, ...]` so that's fine. Thanks for the comment!

File: src/gt4py/stencil_object.py
Comment (jdahm) [reply src/gt4py/stencil_object.py:133]: I like your naming scheme. I just chose whatever came into my head first, but I hadn't really reviewed this PR myself much yet. Thanks!

File: src/gt4py/stencil_object.py
Comment (egparedes) [inline src/gt4py/stencil_object.py:45]: No need to unpack `field_data` tuple since it is hashable.
```suggestion
    return hash((field_data, *parameter_args.keys(), dumps(domain), dumps(origin)))
```

File: tests/test_integration/test_stencil_object.py
Comment (egparedes) [inline tests/test_integration/test_stencil_object.py:52]: It think the `time.perf_counter()` calls for `start_time` and `end_time` variables are not needed here since the same information is available in the `exec_info` dict as `exec_info["call_run_start_time"] ` and `exec_info["call_run_end_time"] `

**Review by egparedes — CHANGES_REQUESTED** [review]
It looks good, but there are some bugs yet. I outlined possible fixes in my review.

File: src/gt4py/stencil_object.py
Comment (jdahm) [reply src/gt4py/stencil_object.py:41]: Thanks for looking into that!

File: tests/test_integration/test_stencil_object.py
Comment (egparedes) [inline tests/test_integration/test_stencil_object.py:33]: I think creating a fixture here is a bit of an overkill, so I suggest to use a parametrize argument:

```suggestion
@pytest.mark.parametrize("backend", ["gtc:numpy"])
def test_stencil_object_cache(backend: str):
```

**Review by egparedes — APPROVED** [review]
It looks good.

**Comment by egparedes** [comment]
bors try

**Review by egparedes — CHANGES_REQUESTED** [review]
This PR is now solid and it could be merged already, but I would like to ask for one last missing feature: let users clean the domain/origin cache on request. This extra feature could be used for reproducibilty or benchmarking purposes, as well as to free up some memory in case the cache grows too large (although it seems unlikely).

I propose to add a new class method (e.g. `clean_call_args_cache()`) to `StencilObject` which will just call to `cls._domain_origin_cache.clear()`

File: tests/test_integration/test_stencil_object.py
Comment (egparedes) [inline tests/test_integration/test_stencil_object.py:61]: Add small test case for cache cleaning method and minor code style changes.

```suggestion
    # When an origin changes, it needs to recompute more, so the time should increase
    other_out_storage = gt_storage.ones(
        backend=backend, default_origin=(1, 0, 0), shape=shape, dtype=float
    )
    other_origin_time = runit(in_storage, other_out_storage, offset=1.0)
    assert other_origin_time > fast_time
    
    # When the cache is cleared everything is recomputed again, so the time will increase
    cleaned_cache_time = runit(in_storage, out_storage, offset=1.0)
    assert cleaned_cache_time > fast_time    
    
```

**Review by egparedes — COMMENTED** [review]
It looks very good. I'll wait for all the tests to pass and I will merge afterwards.

**Review by egparedes — APPROVED** [review]
Good to go.

**Pull request merged**
